orchest/orchest
# Introduce Orchest Kubernetes Operator

## Description

This PR introduces the `orchest-controller` and the CRD that respectively manage and define `orchestclusters` resources. 

Big step towards a better solution for #866 and improving the management of the Orchest Cluster, e.g. installing it.

## Instructions on how to test

#### Deploying the `orchest-controller`
1. Create the Orchest namespace, e.g. `kubectl create ns orchest`. The namespace is required, given that the operator will be deployed in it.
2. Build the controller's `Dockerfile` on the `minikube` node, using:
    ```
    scripts/build_container.sh -v -i orchest-controller -t "latest"
    ``` 
    (It is listed in the below Todo section to migrate away from the `"latest"` tag.)
3. Deploy the `orchest-controller` and CRD:
   ```
   kubectl -f apply services/orchest-controller/deploy-controller
   ```
   
**Deploying an updated `orchest-controller`** (e.g. after changes):
1. Remove the CR Object, which will also remove the corresponding row in etcd, so that we can start from scratch when deploying a new controller. Go to `orchestcluster` in k9s, or using `kubectl`: `kubectl -n orchest delete orchestcluster cluster-1`
2. Build the new `orchest-controller`
3. Remove the `orchest-controller` pod. This will immediately create a new pod (since it is a deployment)

#### Setting up the CLI

1. Install the `orchest-cli` in its own virtualenv
    ```
    python3 -m venv venv && source venv/bin/active
    # Make sure you are in the `orchest-cli/` directory 
    pip install -e .
    ```
2. You should now be able to run commands like `orchest install`.

#### Set up Orchest for testing

You will need to build the minimal set of containers required to run Orchest with the tag `v2022.04.0` (given that that is currently the version that `orchest install` will get you):
```
eval $(minikube docker-env)
scripts/build_container.sh -m -t "v2022.04.0" -o "v2022.04.0"
```

Now you can run `orchest install` followed by any other command you would like to test.

**Testing `orchest update`**:
1. Build the images with another tag (this should be super quick given you have already build them before). I recommend using `v2022.04.1` so you can test update multiple times
2. Run `orchest update --version=v2022.04.1` (otherwise it will get the latest available version from our update-info endpoint)

## Todo
Here follows an extensive list of items that need to be done before we can merge:
- [ ] Merge in `dev` with needed DB migration fixes.
- [ ] Check with `orchest-ctl` whether we have all the functionality we need.
- [ ] GitHub Action workflow files to publish CLI on release.
- [ ] Release checklist similar to SDK to ensure the `_version.py` is updated. Needs to be included
  in the PR template as well.
- [ ] Installation docs (and how to self-host as you might need nginx reverse-proxy)
  - [ ] We will recommend to install Orchest on a clean cluster (a non-existing cluster) because it is hard to play well with other software already installed on the cluster, e.g. ingress, argo, etc.
  - [ ] Open an issue to fix the above point.
  - [ ] Show usage of CLI and direct yaml apply. The controller will get a service and ingress rule
        as well.
- [ ] `update` and `restart` to work in the GUI
  - [ ] `orchest-controller` needs to get service and ingress rules so that its endpoints is reachable from outside the cluster. These files are to be added to the `deploy-controller` directory.
- [ ] Seems to be an issue with `dev` mode? The orchest-api doesn't start correctly and thus the webserver also can't query it for the config settings.
- [ ] Seems like the `cleanup` pod fails?
- [ ] Whenever the CRD is changed, users need to `kubectl apply` its changes. Otherwise the controller might use fields that aren't defined yet. How do we plan to tackle this? Do we want to incorporate anything in the CLI?
- [ ] Removal of `orchest-ctl` from codebase (and required changes to `namespace_ctl.py`)
- [ ] `orchest-controller` needs to be versioned as well instead of using `"latest"` tags. This includes changing the `deployment.yaml` template.
  - [ ] `orchest-controller` to contain the right default version. Otherwise it will only ever install the hardcoded default version of `v2022.04.0`

Items we can postpone for after the release:
- [ ] `uninstall` command to be added. Delete CR Object and then delete `orchest` namespace. Alternatively, we can also tell users to delete their cluster if they were using it solely for Orchest.
- [ ] Development docs around the operator
  - [ ] Note that the `docker-registry` is managed through Helm and not the `orchest-controller`.
  - [ ] How to run the operator outside of the cluster for easy debugging
